roachtest: small improvements in documentation and error reporting

renatolabs opened the pull request:
**roachtest: mixedversion documentation updates**

* comments were added to indicate that callers should not close the
  connection pools returned by helper functions.
* documentation for `InMixedVersion` and `AfterUpgradeFinalized` were
  updated to be more explicit about when the hooks passed to those
  functions can be called.

**roachtest: include step ID in failures to get debug information**

This commit updates the `mixedversion` framework so that we include
the step ID we were preparing to run when an error is found gathering
debug information (current binary and cluster versions on the
nodes). This makes it a little easier to understand these errors and
at what point in th test they happened.

Epic: none

Release note: None

Commit:
roachtest: mixedversion documentation updates

* comments were added to indicate that callers should not close the
  connection pools returned by helper functions.
* documentation for `InMixedVersion` and `AfterUpgradeFinalized` were
  updated to be more explicit about when the hooks passed to those
  functions can be called.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129089)
<!-- Reviewable:end -->

msbutler commented on pkg/cmd/roachtest/roachtestutil/mixedversion/helper.go:
```diff
@@ -72,10 +72,17 @@ type (
 	}
 )
 
+// Connect returns a connection pool to the given node. Note that
```
nit: i think it would be clearer to state that `Connect` returns a connection _from a_ pool, not a "connection pool", right? 

Same comment applies below.

renatolabs replied:
`*sql.DB` is a connection pool. `*sql.Conn` is a single connection. Most existing functions in roachtest (and elsewhere) were written with the assumption that you pass the former. It's also the recommended approach:

> Prefer running queries from [DB](https://pkg.go.dev/database/sql#DB) unless there is a specific need for a continuous single database connection.

https://pkg.go.dev/database/sql#Conn

renatolabs replied:
Heh, true, I think that's common in our codebase in general. Roachtest's `c.Conn()` is the same.

Commit:
roachtest: include step ID in failures to get debug information

This commit updates the `mixedversion` framework so that we include
the step ID we were preparing to run when an error is found gathering
debug information (current binary and cluster versions on the
nodes). This makes it a little easier to understand these errors and
at what point in th test they happened.

Epic: none

Release note: None

renatolabs commented:
Pushed a commit to include the step ID when there's an error gathering debug information (binary and cluster versions on the nodes).

msbutler commented:
While you're here, it would be great to add more details on the calling behavior of `AfterUpgradeFinalized` (called once per major upgrade), and `InMixedVersion` called up to 3 terms per major upgrade (upgrade, downgrade, upgrade again)

renatolabs commented:
> While you're here, it would be great to add more details on the calling behavior of `AfterUpgradeFinalized` (called once per major upgrade), and `InMixedVersion` called up to 3 terms per major upgrade (upgrade, downgrade, upgrade again)

Makes sense, updated the comments for those functions as well! I'm due to revisit a lot of the documentation and README once work on multitenant support is finished, as there's a lot of changes that are not fully covered yet.

renatolabs commented:
TFTR!

bors r=msbutler
